torch/csrc/jit/tensorexpr/ir.cpp

```diff
@@ -10,46 +10,183 @@ static Dtype ChooseDtype(const Dtype& buffer_dtype, const Dtype& index_dtype) {
   }
-
-  if (index->dtype().lanes() != mask->dtype().lanes()) {
+  if (!indicesValid(indices)) {
     throw malformed_input();
   }
-
-  if (index->dtype().scalar_type() != ScalarType::Int) {
-    throw unsupported_dtype();
+  Dtype index_dtype = dtypeOfIndices(indices);
+  if (index_dtype.lanes() != mask->dtype().lanes()) {
+    throw malformed_input();
   }
 }
 
+ExprHandle Load::make(
+    const Buffer& buffer,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& mask) {
+  return ExprHandle(
+      new Load(buffer, ExprHandleVectorToExprVector(indices), mask.node()));
+}
+ExprHandle Load::make(
+    Dtype dtype,
+    const BufHandle& buf,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& mask) {
+  return ExprHandle(new Load(
+      dtype, buf.node(), ExprHandleVectorToExprVector(indices), mask.node()));
+}
+
 Store::Store(
     const Buffer& buffer,
-    const Expr* index,
+    const std::vector<const Expr*>& indices,
     const Expr* value,
     const Expr* mask)
-    : Store(buffer.data(), index, value, mask) {
+    : Store(buffer.data(), indices, value, mask) {
   if (buffer.dtype().scalar_type() != value->dtype().scalar_type()) {
     throw malformed_input();
   }
 }
 
+Store::Store(
+    const Buf* buf,
+    const std::vector<const Expr*>& indices,
+    const Expr* value,
+    const Expr* mask)
+    : buf_(buf), indices_(indices), value_(value), mask_(mask) {
+  if (buf->dtype() != kHandle) {
+    throw malformed_input();
+  }
+  /*
+  TODO: Reenable the checks.
+  The reason they are disabled is that kernel.cpp is using Buffers somewhat
+  loosely: we don't set dimensions properly and just construct index expressions
+  directly. We should harden that part and then we'd be able to turn on these
+  checks.
+
+  if (!indicesValid(indices)) {
+    throw malformed_input();
+  }
+  if (!mask || !value) {
+    throw malformed_input();
+  }
+  Dtype index_dtype = dtypeOfIndices(indices);
+  if (index_dtype.lanes() != mask->dtype().lanes()) {
+    throw malformed_input();
+  }
+  if (index_dtype.lanes() != value->dtype().lanes()) {
+    throw malformed_input();
+  }
+  */
+}
+
+Store* Store::make(
+    const Buffer& buffer,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& value,
+    const ExprHandle& mask) {
+  return new Store(
+      buffer, ExprHandleVectorToExprVector(indices), value.node(), mask.node());
+}
+
+Store* Store::make(
+    const BufHandle& buf,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& value,
+    const ExprHandle& mask) {
+  return new Store(
+      buf.node(),
+      ExprHandleVectorToExprVector(indices),
+      value.node(),
+      mask.node());
+}
+
+Store* Store::make(
+    const BufHandle& buf,
+    const std::vector<ExprHandle>& indices,
+    const ExprHandle& value) {
+  return new Store(
+      buf.node(),
+      ExprHandleVectorToExprVector(indices),
+      value.node(),
+      ExprHandle(1).node());
+}
+
+const Expr* flattened_index(
```

Comment:
Minor: I'd prefer flat_index for shorter name, instead of flattened_index. But it's up to you.

Reply:
Sounds good, I'll rename!